src/Build.UnitTests/Graph/ProjectGraph_Tests.cs

```diff
@@ -842,14 +1094,486 @@ public void TopologicalSortShouldTopologicallySort(Dictionary<int, int[]> edges)
+        {
+            var root = CreateProjectFile(_env, 1, new[] {2}, null, null, OuterBuildSpecification).Path;
+            CreateProjectFile(_env, 2);
+
+            var graph = new ProjectGraph(root);
+
+            var dot = graph.ToDot();
+
+            graph.ProjectNodes.Count.ShouldBe(4);
+
+            AssertOuterBuildAsRoot(graph.GraphRoots.First());
+
+            var nonMultitargetingNode = GetFirstNodeWithProjectNumber(graph, 2);
+
+            AssertNonMultitargetingNode(nonMultitargetingNode);
+        }
+
+        [Fact]
+        public void InnerBuildAsRootViaLocalPropertyShouldNotPropagateInnerBuildPropertyToReference()
+        {
+            var innerBuildViaLocalProperty = OuterBuildSpecification + $"<PropertyGroup><{InnerBuildProperty}>foo</{InnerBuildProperty}></PropertyGroup>";
+
+            var root = CreateProjectFile(
+                _env,
+                1,
+                new[] {2},
+                null,
+                null,
+                innerBuildViaLocalProperty).Path;
+
+            CreateProjectFile(_env, 2);
+
+            var graph = new ProjectGraph(root);
+
+            var dot = graph.ToDot();
+
+            graph.ProjectNodes.Count.ShouldBe(2);
+
+            AssertInnerBuildEvaluation(graph.GraphRoots.First(), false, new Dictionary<string, string>());
+
+            var nonMultitargetingNode = GetFirstNodeWithProjectNumber(graph, 2);
+
+            AssertNonMultitargetingNode(nonMultitargetingNode);
+        }
+
+        [Fact]
+        public void InnerBuildAsRootViaGlobalPropertyShouldNotPropagateInnerBuildPropertyToReference()
+        {
+            var root = CreateProjectFile(_env, 1, new[] {2}, null, null, OuterBuildSpecification).Path;
+            CreateProjectFile(_env, 2);
+
+            var graph = new ProjectGraph(root, new Dictionary<string, string>{{InnerBuildProperty, "foo"}});
+
+            var dot = graph.ToDot();
+
+            graph.ProjectNodes.Count.ShouldBe(2);
+
+            AssertInnerBuildEvaluation(graph.GraphRoots.First(), true, new Dictionary<string, string>());
+
+            var nonMultitargetingNode = GetFirstNodeWithProjectNumber(graph, 2);
+
+            AssertNonMultitargetingNode(nonMultitargetingNode);
+        }
+
+        [Fact]
+        public void NonOuterBuildProjectsInTheMiddle()
+        {
+            var root = CreateProjectFile(_env, 1, new[] {2, 3}, null, null, OuterBuildSpecification).Path;
+            CreateProjectFile(_env, 2, new[] {4});
+            CreateProjectFile(_env, 3, new[] {4});
+            CreateProjectFile(_env, 4, null, null, null, OuterBuildSpecification);
+
+            var graph = new ProjectGraph(root);
+
+            var dot = graph.ToDot();
+
+            graph.ProjectNodes.Count.ShouldBe(8);
+
+            AssertOuterBuildAsRoot(graph.GraphRoots.First());
+            AssertOuterBuildAsNonRoot(GetNodesWithProjectNumber(graph, 4).First(IsOuterBuild));
```

Comment:
Also assert 2 inner builds of p4?

Reply:
Already asserted in the two `AssertOuterBuildAs*` methods.